### PR TITLE
Cross subnet IPIP

### LINKED ISSUE #1127: Ability to use ipip only when communicating across different subnets
Currently enabling ipip will force every connection to use ipip encapsulation. It would however be more optimal to let Calico to use native connection whenever both src and dst hosts are in the same network, but fallback to ipip when they are in different networks and when the cross network router doesn't support BGP (thus as in AWS EC2 VPC). This would give the best possible performance in both inter-subnet traffic and cross-subnet traffic.

After some discussion in the IRC channel I got the impression that this might be possible relatively easy by modifying the bird config.

### PR DESCRIPTION
**Pull request by robbrockbank**
Fixes #1127 

Replaces #1453 (PR got messed up)

This PR applies the minimal feedback from #1453 (see closed out PR for details), but also adds STs.

Documentation covered by https://github.com/projectcalico/calico/pull/444

### CODE REVIEW

File: tests/st/calicoctl/test_ipip.py [hunk 1]
```diff
@@ -130,3 +203,53 @@ def remove_tunl_ip(self):
...
+        Test whether IPIP is being used as expected on host1 when pinging workload_host2
+        from workload_host1.
+        """
+        def check():
+            orig_tx = self.get_tunl_tx(host1)
+            workload_host1.execute("ping -c 2 -W 1 %s" % workload_host2.ip)
+            if expect_ipip:
+                assert self.get_tunl_tx(host1) == orig_tx + 2
+            else:
+                assert self.get_tunl_tx(host1) == orig_tx
```
Comment (gunjan5) [inline tests/st/calicoctl/test_ipip.py]: Where is this function asserting if the ping succeeds when `expect_ipip` is `False`? Only assertion being made here is tunl TX count is the same which would be the case even if ping doesn't work.

File: tests/st/utils/docker_host.py [hunk 2]
```diff
@@ -6,12 +6,13 @@
 #
 #     http://www.apache.org/licenses/LICENSE-2.0
 #
-# Unless required by applicable law or agreed to in writing, software
+# Unless required by appliccable law or agreed to in writing, software
```
Comment (gunjan5) [inline tests/st/utils/docker_host.py]: apliccable -> applicable

File: calico_node/filesystem/etc/calico/confd/templates/bird_ipam.cfg.template [hunk 3]
```diff
@@ -16,15 +16,14 @@ filter calico_ipip {
   if ( net ~ {{$data.cidr}} ) then {
 {{if eq $data.ipip_mode "cross-subnet"}}
     if ( from ~ {{$network}} ) then
-      krt_tunnel = "";
+      krt_tunnel = "";                     {{/* Destination in pool, mode is cross sub-net, route from-host on subnet, do not use IPIP */}}
```
Comment (gunjan5) [inline calico_node/filesystem/etc/calico/confd/templates/bird_ipam.cfg.template]: nit: pool -> ipPool 😛

File: tests/st/calicoctl/test_ipip.py [hunk 4]
```diff
@@ -130,3 +203,53 @@ def remove_tunl_ip(self):
...
+        return current_prefix_len
+
+    def assert_ipip_routing(self, host1, workload_host1, workload_host2, expect_ipip):
+        """
+        Test whether IPIP is being used as expected on host1 when pinging workload_host2
+        from workload_host1.
+        """
+        def check():
+            orig_tx = self.get_tunl_tx(host1)
+            workload_host1.execute("ping -c 2 -W 1 %s" % workload_host2.ip)
```
Comment (robbrockbank) [inline tests/st/calicoctl/test_ipip.py]: That's why it's in a retry_until_success loop thing

File: tests/st/calicoctl/test_ipip.py (hunk 1 above)
Comment (robbrockbank) [inline tests/st/calicoctl/test_ipip.py]: If ping doesn't work, the ping will raise a command exec error

File: tests/st/bgp/test_update_ip_addr.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+            self.assert_ip_connectivity(workload_list=[workload_host1,
+                                                       workload_host2],
+                                        ip_pass_list=[workload_host1.ip,
+                                                      workload_host2.ip])
+
+    def _fix_ip(self, host):
+        """
+        Update the calico node resource to have the correct IP for the host.
+        """
+        noder = json.loads(host.calicoctl(
```
Comment (robbrockbank) [inline tests/st/bgp/test_update_ip_addr.py]: not a typo - i meant it as node resource.   I'm not sure why I didn't think node was fine as it was.   Yeah I might change that.

File: tests/st/utils/docker_host.py (hunk 2 above)
Comment (robbrockbank) [inline tests/st/utils/docker_host.py]: Odd, how did that creep in there.

File: tests/st/calicoctl/test_ipip.py [hunk 6]
```diff
@@ -130,3 +203,53 @@ def remove_tunl_ip(self):
...
+            if expect_ipip:
+                assert self.get_tunl_tx(host1) == orig_tx + 2
+            else:
+                assert self.get_tunl_tx(host1) == orig_tx
+        retry_until_success(check)
```
Comment (robbrockbank) [inline tests/st/calicoctl/test_ipip.py]: yes it does timeout.

File: tests/st/bgp/test_update_ip_addr.py (hunk 5 above)
Comment (gunjan5) [inline tests/st/bgp/test_update_ip_addr.py]: that's ok. Not a big deal

**Pull request merged**